File: app/src/main/java/io/neft/Renderer/Text.java
package io.neft.Renderer;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.FontMetrics;
import android.graphics.Typeface;
import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;

import io.neft.Client.Action;
import io.neft.Client.InAction;
import io.neft.Client.OutAction;
import io.neft.Client.Reader;
import io.neft.MainActivity;
import io.neft.Utils.ColorUtils;

public class Text extends Item {
    static void register(final MainActivity app){
        app.client.actions.put(InAction.CREATE_TEXT, new Action() {
            @Override
            public void work(Reader reader) {
                new Text(app);
            }
        });

        app.client.actions.put(InAction.SET_TEXT, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setText(reader.getString());
            }
        });

        app.client.actions.put(InAction.SET_TEXT_WRAP, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setWrap(reader.getBoolean());
            }
        });

        app.client.actions.put(InAction.UPDATE_TEXT_CONTENT_SIZE, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).updateContentSize();
            }
        });

        app.client.actions.put(InAction.SET_TEXT_COLOR, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setColor(reader.getInteger());
            }
        });

        app.client.actions.put(InAction.SET_TEXT_LINE_HEIGHT, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setLineHeight(reader.getFloat());
            }
        });

        app.client.actions.put(InAction.SET_TEXT_FONT_FAMILY, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setFontFamily(reader.getString());
            }
        });

        app.client.actions.put(InAction.SET_TEXT_FONT_PIXEL_SIZE, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setFontPixelSize(reader.getFloat());
            }
        });

        app.client.actions.put(InAction.SET_TEXT_FONT_WORD_SPACING, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setFontWordSpacing(reader.getFloat());
            }
        });

        app.client.actions.put(InAction.SET_TEXT_FONT_LETTER_SPACING, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setFontLetterSpacing(reader.getFloat());
            }
        });

        app.client.actions.put(InAction.SET_TEXT_ALIGNMENT_HORIZONTAL, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setAlignmentHorizontal(reader.getString());
            }
        });

        app.client.actions.put(InAction.SET_TEXT_ALIGNMENT_VERTICAL, new Action() {
            @Override
            public void work(Reader reader) {
                ((Text) app.renderer.getItemFromReader(reader)).setAlignmentVertical(reader.getString());
            }
        });

        app.client.actions.put(InAction.LOAD_FONT, new Action() {
            @Override
            public void work(Reader reader) {
                Text.loadFont(reader.getString(), reader.getString(), app);
            }
        });
    }

    private class Line {
        public int textStart;
        public int textEnd;
        public float width;
        public float y;
    }

    private static final FontMetrics fontMetrics = new FontMetrics();

    private static final HashMap<String, Float> alignment;
    static {
        alignment = new HashMap<>();
        alignment.put("top", 0f);
        alignment.put("center", 0.5f);
        alignment.put("bottom", 1f);
        alignment.put("left", 0f);
        alignment.put("right", 1f);
    };

    static final Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    static final HashMap<String, Typeface> fonts = new HashMap<>();

    public String text = "";
    public int color = Color.BLACK;
    public boolean wrap = false;
    public float lineHeight = 1;
    public String fontFamily = "";
    public float fontPixelSize;
    public float contentWidth = 0;
    public float contentHeight = 0;
    public float alignmentHorizontal = alignment.get("left");
    public float alignmentVertical = alignment.get("top");

    protected String linesText = "";
    protected ArrayList<Line> lines;
    protected int linesLength;
    protected String[] words;

    public static void loadFont(final String name, final String source, final MainActivity app){
        final Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                if (source.startsWith("/static")){
                    fonts.put(name, Typeface.createFromAsset(app.getAssets(), source.substring(1)));
                    app.client.pushAction(OutAction.FONT_LOAD);
                    app.client.pushString(name);
                    app.client.pushBoolean(true);
                } else {
                    Log.e("JAVA", "Loading font's by URL is not currently supported");
                    app.client.pushAction(OutAction.FONT_LOAD);
                    app.client.pushString(name);
                    app.client.pushBoolean(false);
                }
            }
        });

        thread.start();
    }

    public Text(MainActivity app){
        super(app);
        this.lines = new ArrayList<>();
        this.fontPixelSize = app.renderer.dpToPx(14);
    }

    public void setText(String val){
        text = val;
        words = text.split(" ");
    }

    public void setWrap(boolean val){
        wrap = val;
    }

    public void setColor(int val){
        color = ColorUtils.RGBAtoARGB(val);
        invalidate();
    }

    public void setLineHeight(float val){
        lineHeight = val;
    }

    public void setFontFamily(String val){
        fontFamily = val;
    }

    public void setFontPixelSize(float val){
        fontPixelSize = app.renderer.dpToPx(val);
    }

    // TODO
    public void setFontWordSpacing(float val){

    }

    // TODO
    public void setFontLetterSpacing(float val){

    }

    public void setAlignmentHorizontal(String val){
        alignmentHorizontal = alignment.get(val);
        invalidate();
    }

    public void setAlignmentVertical(String val){
        alignmentVertical = alignment.get(val);
        invalidate();
    }

    public void updateContentSize(){
        linesLength = 0;
        linesText = text;

        if (words != null){
            updatePaint();
            paint.getFontMetrics(fontMetrics);

            final int linesArrayLength = lines.size();
            final int wordsLength = words.length;
            final float lineHeightPx = lineHeight * fontPixelSize;
            final float maxWidth = this.width;
            final float spaceWidth = paint.measureText(" ");
            final float fontTop = lineHeightPx - fontMetrics.descent;
            float x = 0, width = 0;
            float height = 0;
            int textPointer = 0, lineStart = 0, lineLength = 0;

            for (int i = 0; i <= wordsLength; i++){
                final String word = i < wordsLength ? words[i] : null;
                final int wordLength = word != null ? word.length() : 0;
                final float wordWidth = word != null ? paint.measureText(word) : 0;

                if (i > 0 && (i == wordsLength || (wrap && x + wordWidth > maxWidth))){
                    // get line object
                    final Line line;
                    if (linesArrayLength <= linesLength){
                        line = new Line();
                        lines.add(line);
                    } else {
                        line = lines.get(linesLength);
                    }
                    linesLength++;

                    // fill line
                    line.textStart = lineStart;
                    line.textEnd = lineStart + lineLength - 1;
                    line.width = x - spaceWidth;
                    line.y = height + fontTop;

                    // clear
                    lineStart = textPointer;
                    lineLength = 0;
                    x = 0;

                    // next line
                    height += lineHeightPx;
                }

                x += wordWidth + spaceWidth;
                textPointer += wordLength + 1;
                lineLength += wordLength + 1;

                if (x > width){
                    width = x;
                }
            }

            contentWidth = width;
            contentHeight = height;
        } else {
            contentWidth = contentHeight = 0;
        }

        invalidate();

        // push data
        app.client.pushAction(OutAction.TEXT_SIZE);
        app.renderer.pushItem(this);
        app.client.pushFloat(app.renderer.pxToDp(contentWidth));
        app.client.pushFloat(app.renderer.pxToDp(contentHeight));
    }

    protected void updatePaint(){
        final Typeface font = fonts.get(fontFamily);
        if (font != null) {
            paint.setTypeface(font);
        } else {
            paint.setTypeface(Typeface.create(Typeface.SANS_SERIF, Typeface.NORMAL));
        }

        paint.setTextSize(fontPixelSize);
    }

    @Override
    protected void drawShape(final Canvas canvas, final int alpha){
        if (contentWidth <= 0 || contentHeight <= 0){
            return;
        }

        updatePaint();
        paint.setColor(color);
        paint.setAlpha(alpha);

        final float shiftY = (this.height - this.contentHeight) * alignmentVertical;

        for (int i = 0; i < linesLength; i++){
            final Line line = lines.get(i);

            canvas.drawText(linesText,
                    line.textStart, line.textEnd,
                    (this.width - line.width) * alignmentHorizontal, line.y + shiftY,
                    paint);
        }
    }
}

File: Neft/Text.swift
import SpriteKit

extension Renderer {
    class Text: Renderer.BaseType {
        class Node: Item.Node {
            static var fontAttrs: [String: NSObject] = [NSFontAttributeName: NSObject()]
            static let wordBreaks = NSCharacterSet(charactersInString: "- ")
            
            class Label: SKLabelNode {
                var width: CGFloat = 0
                var y: CGFloat = 0
            }
            
            let content = SKNode()
            var labels: [Node.Label] = []
            var labelsLength = 0
            var font: UIFont = UIFont(name: "Helvetica", size: 14)!
            var contentWidth: CGFloat = 0
            var contentHeight: CGFloat = 0
            var text: NSString = ""
            var wrap = false
            var color: UIColor = UIColor.blackColor()
            var lineHeight: CGFloat = 1
            var fontFamily = "Helvetica"
            var fontPixelSize: CGFloat = 14
            var alignmentHorizontal = "left"
            var alignmentVertical = "top"
            
            override init(_ renderer: Renderer) {
                super.init(renderer)
                self.childrenNode.addChild(self.content)
            }

            required init?(coder aDecoder: NSCoder) {
                fatalError("init(coder:) has not been implemented")
            }
            
            private func getLabel(index: Int) -> Label {
                var label: Label? = nil
                if self.labelsLength > index {
                    label = self.labels[index]
                } else {
                    label = Label()
                    labels.append(label!)
                    label!.horizontalAlignmentMode = .Left
                    label!.verticalAlignmentMode = .Top
                }
                return label!
            }
            
            private func updateLabelFont(label: Label) {
                label.fontColor = color
                label.fontName = fontFamily
                label.fontSize = fontPixelSize
            }
            
            func updateContent() {
                if (self.text.length == 0){
                    self.contentWidth = 0
                    self.contentHeight = 0
                    self.labelsLength = 0
                    self.content.removeAllChildren()
                    return
                }
                
                Node.fontAttrs[NSFontAttributeName] = self.font
                
                let fontAttrs = Node.fontAttrs
                let wordBreaks = Node.wordBreaks
                var labelsLength = 0
                var width: CGFloat = 0
                let lineHeight = fontPixelSize * self.lineHeight
                var height: CGFloat = 0
                let maxWidth: CGFloat? = self.wrap ? self.width : nil
                let text = self.text

                // by new lines
                var lineNextString = text
                while lineNextString.length > 0 {
                    var lineText: NSString = ""
                    let newLineIndex = lineNextString.rangeOfString("\n").location
                    if newLineIndex != NSNotFound {
                        lineText = lineNextString.substringToIndex(newLineIndex)
                        lineNextString = lineNextString.substringFromIndex(newLineIndex + 1)
                    } else {
                        lineText = lineNextString
                        lineNextString = ""
                    }
                    
                    // by words
                    var x: CGFloat = 0
                    var wrapLineLength = 0
                    var wordsNextString = lineText
                    while wordsNextString.length > 0 {
                        var wordText: NSString = ""
                        let spaceIndex = wordsNextString.rangeOfCharacterFromSet(wordBreaks).location
                        if spaceIndex != NSNotFound {
                            wordText = wordsNextString.substringToIndex(spaceIndex)
                            wordsNextString = wordsNextString.substringFromIndex(spaceIndex + 1)
                        } else {
                            wordText = wordsNextString
                            wordsNextString = ""
                        }
                        
                        // get size
                        let wordWidth = wordText.sizeWithAttributes(fontAttrs).width
                        let breakLine = wordsNextString != "" && maxWidth != nil && x + wordWidth > maxWidth
                        
                        if wordsNextString == "" || !breakLine {
                            x += wordWidth
                            wrapLineLength += wordText.length
                        }
                        
                        if wordsNextString == "" || breakLine {
                            let label = self.getLabel(labelsLength)
                            label.y = height - lineHeight
                            label.width = x
                            label.text = lineText.substringToIndex(wrapLineLength)
                            lineText = lineText.substringFromIndex(wrapLineLength)
                            if label.parent == nil {
                                self.content.addChild(label)
                            }
                            self.updateLabelFont(label)
                            labelsLength += 1
                            height += lineHeight
                            
                            if breakLine {
                                x = wordWidth
                                wrapLineLength = wordText.length
                            }
                        }
                        
                        if x > width {
                            width = x
                        }
                        
                        wrapLineLength += 1
                    }
                }
                
                // align labels
                var shiftY = self.height - height
                switch self.alignmentVertical {
                case "center":
                    shiftY /= 2
                case "bottom":
                    break
                default:
                    shiftY = 0
                }
                for i in 0..<labelsLength {
                    let label = self.labels[i]
                    
                    // horizontal
                    let x = self.width - label.width
                    switch self.alignmentHorizontal {
                    case "center":
                        label.position.x = x / 2
                    case "right":
                        label.position.x = x
                    default:
                        label.position.x = 0
                    }
                    
                    // vertical
                    label.position.y = -label.y - shiftY
                }
                
                // remove unused labels
                if self.labelsLength > labelsLength {
                    for i in labelsLength..<self.labelsLength {
                        let label = self.labels[i]
                        label.removeFromParent()
                    }
                }
                
                self.labelsLength = labelsLength
                self.contentWidth = width
                self.contentHeight = height
            }
        }
        
        override init(app: GameViewController) {
            super.init(app: app)
            Renderer.actions[InActions.CREATE_TEXT] = self.create
            Renderer.actions[InActions.SET_TEXT] = self.setText
            Renderer.actions[InActions.SET_TEXT_WRAP] = self.setWrap
            Renderer.actions[InActions.UPDATE_TEXT_CONTENT_SIZE] = self.updateContentSize
            Renderer.actions[InActions.SET_TEXT_COLOR] = self.setColor
            Renderer.actions[InActions.SET_TEXT_LINE_HEIGHT] = self.setLineHeight
            Renderer.actions[InActions.SET_TEXT_FONT_FAMILY] = self.setFontFamily
            Renderer.actions[InActions.SET_TEXT_FONT_PIXEL_SIZE] = self.setFontPixelSize
            Renderer.actions[InActions.SET_TEXT_FONT_WORD_SPACING] = self.setFontWordSpacing
            Renderer.actions[InActions.SET_TEXT_FONT_LETTER_SPACING] = self.setFontLetterSpacing
            Renderer.actions[InActions.SET_TEXT_ALIGNMENT_HORIZONTAL] = self.setAlignmentHorizontal
            Renderer.actions[InActions.SET_TEXT_ALIGNMENT_VERTICAL] = self.setAlignmentVertical
            Renderer.actions[InActions.LOAD_FONT] = self.loadFont
        }
        
        private func pushContentSize(item: Node) {
            renderer.pushAction(OutActions.TEXT_SIZE)
            renderer.pushInteger(item.id)
            renderer.pushFloat(item.contentWidth)
            renderer.pushFloat(item.contentHeight)
        }
        
        func create(reader: Reader) {
            Node(renderer)
        }
        
        func setText(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getString()
            item.text = val
            item.updateContent()
            pushContentSize(item)
        }
        
        func setWrap(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getBoolean()
            item.wrap = val
            item.updateContent()
            pushContentSize(item)
        }
        
        func updateContentSize(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            item.updateContent()
            pushContentSize(item)
        }
        
        func setColor(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getInteger()
            let color = hexColorToUIColor(val)
            item.color = color
            
            for i in 0..<item.labelsLength {
                let label = item.labels[i]
                label.fontColor = color
            }
        }
        
        func setLineHeight(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getFloat()
            item.lineHeight = val
            item.updateContent()
            pushContentSize(item)
        }
        
        func setFontFamily(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getString()
//            item.fontFamily = val
//            item.font = UIFont(name: val, size: item.fontPixelSize)!
            item.updateContent()
            pushContentSize(item)
        }
        
        func setFontPixelSize(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getFloat()
            item.fontPixelSize = val
            item.font = UIFont(name: item.fontFamily, size: val)!
            item.updateContent()
            pushContentSize(item)
        }
        
        func setFontWordSpacing(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getFloat()
            // TODO
        }
        
        func setFontLetterSpacing(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getFloat()
            // TODO
        }
        
        func setAlignmentHorizontal(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getString()
            item.alignmentHorizontal = val
            item.updateContent()
        }
        
        func setAlignmentVertical(reader: Reader) {
            let item = renderer.getItem(reader) as! Node
            let val = reader.getString()
            item.alignmentVertical = val
            item.updateContent()
        }
        
        func loadFont(reader: Reader) {
            let name = reader.getString()
            let source = reader.getString()
            // TODO
        }
    }
}
